config/config_test.go

```diff
@@ -0,0 +1,141 @@
+// Package config provides interfaces and utilities for different Cluster
+// components to register, read, write and validate configuration sections
+// stored in a central configuration file.
+package config_test
+
+import (
+	"bytes"
+	"encoding/json"
+	"reflect"
+	"testing"
+
+	ipfscluster "github.com/ipfs/ipfs-cluster"
+	"github.com/ipfs/ipfs-cluster/api/ipfsproxy"
+	"github.com/ipfs/ipfs-cluster/api/rest"
+	"github.com/ipfs/ipfs-cluster/config"
+	"github.com/ipfs/ipfs-cluster/pintracker/maptracker"
+)
+
+func setupConfigManager() *config.Manager {
+	cfg := config.NewManager()
+	clusterCfg := &ipfscluster.Config{}
+	apiCfg := &rest.Config{}
+	ipfsproxyCfg := &ipfsproxy.Config{}
+	maptrackerCfg := &maptracker.Config{}
+	cfg.RegisterComponent(config.Cluster, clusterCfg)
+	cfg.RegisterComponent(config.API, apiCfg)
+	cfg.RegisterComponent(config.API, ipfsproxyCfg)
+	cfg.RegisterComponent(config.PinTracker, maptrackerCfg)
+	return cfg
+}
+
+func TestManager_ToJSON(t *testing.T) {
+	var want1 bytes.Buffer
+	json.Compact(&want1, []byte(`{
```

Comment:
Overkill? This just needs to test that sections are not empty. Otherwise it's stepping into component configurations (components are responsible their ToJSON() work properly). This module should be fully agnostic to component configurations (if ever we split into multiple repos we'll have a nice circular dep here).

Preferably it should be testing with a mock configuration implementation.

And I really don't want to have to change this every time the configuration changes somewhere.

I also disagree with using reflect (I don't fully understand why, having https://golang.org/pkg/bytes/#Equal).

Sorry for the work here, but overly complicated imho.

Ideally:

* Create mock config which ToJSONs() to something short like `{"a":"b"}`
* Register that to all SectionTypes() on a Config.
* Call ToJSON()
* Do the bytes equal (I guess) on the whole of it (unless ToJson comes unordered, in which case you need to extract section as now).

Not sure we need iterative testcases here (there is no different ways of behaving for ToJSON()), so I only see 1 testcase.

Reply:
I use https://github.com/cweill/gotests to generate the test boilerplate, the default for the tool is to do table-driven and use reflect. I have no issue with changing from DeepEqual to bytes.Equal. I was mainly just trying to get some tests to confirm the fix.